File: solace/cmd/deprecated/play/Scan.java
package solace.cmd.deprecated.play;

import java.util.*;
import solace.game.*;
import solace.net.*;

/**
 * Scan command, shows players and mobiles in adjacent rooms.
 * @author Ryan Sandor Richards
 */
public class Scan extends PlayStateCommand {
  public Scan(solace.game.Character ch) {
    super("scan", ch);
  }

  public void run(Connection c, String []params) {
    if (character.isRestingOrSitting()) {
      character.sendln("You cannot scan around unless standing.");
      return;
    }

    if (character.isFighting()) {
      character.sendln("You are too busy to scan while engaged in combat!");
      return;
    }

    if (character.isSleeping()) {
      character.sendln("You cannot scan whilst asleep.");
      return;
    }

    Room room = character.getRoom();
    Area area = room.getArea();
    List<Exit> exits = room.getExits();
    String message = "";

    for (Exit ex : exits) {
      Room r = area.getRoom(ex.getToId());

      if (r.getCharacters().size() == 0)
        continue;

      message += ex.getDescription().trim() + ":\n\r";

      synchronized(r.getCharacters()) {
        for (Player ch : r.getCharacters()) {
          message += "  " + ch.getName() + "\n\r";
        }
      }
    }

    if (message.equals("")) {
      message = "There is nobody of interest in any direction.";
    }

    // Tell the other players in the room what the character is doing
    room.sendMessage(
      character.getName() + " scans the surrounding area.",
      character
    );

    c.sendln(message);
  }
}

File: game/commands/view/scan.js
'use strict';

/**
 * Scan command, shows players and mobiles in adjacent rooms.
 * @author Ryan Sandor Richards
 */
Commands.add('scan', function (player, params) {
  if (player.isRestingOrSitting()) {
    return player.sendln("You cannot scan around unless standing.");
  }

  if (player.isFighting()) {
    return player.sendln("You are too busy to scan while engaged in combat!");
  }

  if (player.isSleeping()) {
    return player.sendln("You cannot scan whilst asleep.");
  }

  var room = player.getRoom();
  var area = room.getArea();
  var exits = room.getExits();
  var builder = new StringBuilder();

  exits.forEach(function (exit) {
    var r = area.getRoom(exit.getToId());
    if (r.getPlayers().size() == 0) return;
    builder.append(exit.getDescription().trim());
    builder.append(":\n\r");
    r.getPlayers().forEach(function (p) {
      builder.append("  ");
      builder.append(p.getName());
      builder.append("\n\r");
    });
  });

  room.sendMessage(
    format('%s scans the surrounding area.', player.getName()), player);
  var message = builder.toString();
  if (message == "") {
    message = "There is nobody of interest in any direction.";
  }
  player.sendln(message);
});
